### ISSUE TITLE
NKAI dimension door

### ISSUE THREAD

**Issue by nullkiller** [vcmi/vcmi#1599]
**Describe your proposal**
Lets fix some thoughts regarding dimension door and why it is not implemented yet in NKAI. 

AIPathfinder works using kind of Dijkstra's algorithm. Each tile node now has 8 possible neighbors, 8 ways to move further. Now the easiest way to implement DD will be to add all possible casts to these 8 neighbors and let it go. And DD radius is 9 tiles so it is 19^2 additional nodes. Quite huge amount so it can easily kill pathfinder performance already quite slow. So my considerations here is to somehow reduce the amount of nodes without loosing too much. For instance we can calculate only nodes on perimeter of DD and even here not all of them. For instance this way
![image](https://user-images.githubusercontent.com/5644734/220416545-4f2868af-f07d-4bab-83e4-5d42ee7908c6.png)
In this case we have only 24 additional nodes so more or less bearable. We will not be able to make short DD and in some cases it will block some ways but in overall may be more or less ok. Also need to take into account that if we can not rich some place from one tile we still have chance to rich it from adjacent tile.

So the idea is to use DD on every tile and add these 24 new neighbors but switching to casting actor (see how water walking works now) so that we have both regular path and DD path and can choose.
Now if we have more than 1 DD available (e.g. expert air) we need a way to calculate how many we already used. For this we can traverse back path and count how many SpecialAction of type DD we have used this day. For this we will need some way to compare SpecialAction. While this is more or less clear I would better start with no more than 1 DD per day and see how it will perform.

Another options is to use DD only when we encountered some obstacle. It will be harder to implement so.

**Variety of cases**
DD has some interesting options on how it can influence path calculations. For instance we may prefer blinking right away because we are in swamp right now or vise versa we are on grass and want to walk first and blink only when we enter swamp to gain maximum distance. Same with obstacles: some obstacles may be better to go around and some to use DD to bypass. Or we may use all DD before we rich some obstacle we can not go around and now we need DD but we have none already. These cases show that any attempt to fix performance of DD by skipping some obviously ineffective at first glance moves may end up cutting some edge cases and investigating some complex bugs.

**Comment by IvanSavenko** [comment]
Perhaps first consider Town Portal? (unless AI already uses it)
Reasoning: 
- less load on pathfinder. Even with expert Earth number of new nodes for pathfinder = number of towns
- can help with (apparently non-existing/broken) town defense logic.
- can help with army gathering
- less chance that AI traps itself. for example by DD'ing someplace from where he can't go back without another DD (which might require more mana than hero has or waiting for next turn)

**Comment by nullkiller** [comment]
These tickets are not for implementation right now. TP is supported at certain extent.

**Comment by DjWarmonger** [comment]
There is an alternative solution - use DD only to reach areas that are not reachable otherwise. That is, teleporting over zone border or over a strong guard.
Or maybe consider DD only beyond some minimal distance - if it saves a significant amount of movement points.

**Comment by nullkiller** [comment]
Yes, also possible. Probably it can be done when we switch actor. If current one can cast DD we can check all unreachable tiles (or only tiles which are adjacent to blocked tiles) and find best place to cast DD from to rich them and once we created new nodes run pathfinder for them as well

**Comment by krs0** [comment]
According to https://heroes.thelazy.net/index.php/Dimension_Door DD is not symmetrical: (although in VCMI it should be)
_The radius of the spell is 8 squares in top direction and 9 squares in any other direction, diagonal movement included._

![image](https://user-images.githubusercontent.com/15194321/221992896-2c1fb72c-d765-49df-9d9e-3ea69c7dfda1.png)

**Comment by IvanSavenko** [comment]
But if you actually make screenshot in H3 you'll see that this is another bug with DD.
You can only see 8 tiles below hero. 9th tile is (largely) hidden, there is only tiny row which you can use for DD
![зображення](https://user-images.githubusercontent.com/1576820/221994311-a10c0900-fcd4-4a33-9a19-5492573c4c46.png)

**Comment by dydzio0614** [comment]
what about left/right? also pixel hunting to get to 9th row?

**Comment by IvanSavenko** [comment]
No. To left & right H3 actually does have 9 full/almost full tiles, no pixel-hunting required

**Comment by dydzio0614** [comment]
I would say not symmetrical is probably desired, even if it's dumb...
At least for multiplayer reasons people got used to 9 tiles at bottom, and bunch of other people either play with hd mod + high resolution or found out it's possible to get to additional row by pixel hunt (not that hard to eventually find out if 8th tile is blocked but 9th is free)
Implementing this differently may also break some complicated H3 maps where you have to use DD

**Comment by krs0** [comment]
I would not lose any sleep in making this symmetrical. Easier for everyone, especially since you have the DD range rectangle.

**Comment by dydzio0614** [comment]
Some pvp players stated opnion that 1 extra row in teh bottom is good compensation for buildings not having entrance at top...
We will need configurable DD anyway, because of the way this spell works in hota, which is generally favored change by playerbase
